frame/balances/e2e-tests/src/lib.rs

```diff
@@ -0,0 +1,148 @@
+use substrate_test_runner::{Node, ChainInfo};
+use pallet_sudo::Call as SudoCall;
+use pallet_balances::Call as BalancesCall;
+use sp_keyring::sr25519::Keyring::{Alice, Bob};
+use sp_runtime::{traits::IdentifyAccount, MultiSigner, MultiAddress, MultiSignature};
+use sp_runtime::AccountId32;
+use sp_runtime::traits::Block as BlockT;
+use sp_runtime::generic::UncheckedExtrinsic;
+use codec::Encode;
+use frame_support::traits::Get;
+
+type Extrinsic<R, S> = UncheckedExtrinsic<
+    MultiAddress<<R as frame_system::Config>::AccountId, <R as frame_system::Config>::Index>,
+    <R as frame_system::Config>::Call,
+    MultiSignature,
+    S,
+>;
+
+/// Tests the Balances::force_transfer call
+pub fn force_transfer<T>(node: &Node<T>)
+    where
+        T: ChainInfo,
+        T::Runtime: pallet_sudo::Config + pallet_balances::Config,
+        <T::Runtime as frame_system::Config>::Call: From<SudoCall<T::Runtime>> + From<BalancesCall<T::Runtime>> + Encode,
+        <T::Runtime as pallet_sudo::Config>::Call: From<BalancesCall<T::Runtime>>,
+        <T::Runtime as pallet_balances::Config>::Balance: From<u8>,
+        <T::Runtime as frame_system::Config>::AccountId: From<AccountId32> + Encode,
+        <T::Block as BlockT>::Extrinsic: From<Extrinsic<T::Runtime, T::SignedExtras>>,
+        <<T::Runtime as frame_system::Config>::Lookup as sp_runtime::traits::StaticLookup>::Source: From<AccountId32>,
+        AccountId32: std::borrow::Borrow<<T::Runtime as frame_system::Config>::AccountId>
+{
+    type Balances<T> = pallet_balances::Module<T>;
+    let (alice, bob) = (
+        MultiSigner::from(Alice.public()).into_account(),
+        MultiSigner::from(Bob.public()).into_account(),
+    );
+    let (alice_balance, bob_balance) = node.with_state(|| (
+        Balances::<T::Runtime>::free_balance(alice.clone()),
+        Balances::<T::Runtime>::free_balance(bob.clone()),
+    ));
+    let balance = alice_balance / 2u8.into();
+
+    let balances_call = BalancesCall::force_transfer(alice.clone().into(), bob.clone().into(), balance.into());
+    node.submit_extrinsic(
+        SudoCall::sudo(Box::new(balances_call.into())),
+        alice.into()
+    );
```

Comment:
So, in all three tests, I see a few things that makes me feel these tests are not generic. Because:
1) they use Sudo - which outside of dev-chains _shouldn't_ exists and b) didn't we patch the node signature-functions so we can call everything as anyone, not requiring sudo? Why is there Sudo involved? Sudo doesn't exist on any of the final target systems we have (Polkadot, Roccocco, etc...)
2) they use the alice and bob keychains, which only exist on dev-chains. If the target goal is to have a chain provide the actual data and we interact with that, then there _can't  be any alice nor bob_ on it - as they are known. Thus, for any _real data_ these tests are all moot, as neither has any free balance.
3) they only test it into one direction, not resetting the state after. Meaning that the order suddenly matters as one tests sets the balance, while others transfer parts of them. I am not 100% sure this is an actual problem in this case, but generally, I'd expect these tests to transfer one side, assert, and then transfer back and assert, so that we are roughly are where we were before - state-wise.

Reply:
> they use Sudo 

the nature of the `force_transfer` and `set_balance` calls is what requires sudo, we could add a method to `ChainInfo` that takes a call that needs to be dispatched by root, and downstream chains can implement whatever method suits their chain.

> they use the alice and bob keychains

chose alice and bob's chains for simplicity, but before this is merged, these test fn would have to take user supplied accountId's as arguments

> they only test it into one direction, not resetting the state after

yeah this can be trivially fixed by resetting blocks after the test.

Reply:
> chose alice and bob's chains for simplicity, but before this is merged, these test fn would have to take user supplied accountId's as arguments

Are we not in the process of final-review? When should this happen and why not now?

>  yeah this can be trivially fixed by resetting blocks after the test.

So... are you doing that? I am not sure, I understand the statement. I clearly want this to be the case before I agree to merge this, so...